Provide features to handle external (server) errors

danilminibaev opened the pull request.

Commit:
Adding handling validation errors from server feature

Nothing new added here, just because of auto-formatting

Commit:
Adding handling validation errors from server feature

- Provided new property for the server validation errors
- Added merge server side with client side errors on submit to be able
to show whole error set we have at the time

Commit:
Adding handling validation errors from server feature

- Provided property function to be able to toggle some action
(notification, for instance) in case if validation property has
no corresponding field

Commit:
Adding handling validation errors from server feature

- packages updates
- Test are written for From.jsx
- minor fixes in Form.jsx (in componentWillReceiveProps, case when
values are different but we still have server errors)

Commit:
Merge remote-tracking branch 'upstream/master' into feature/server-side-validation-check

# Conflicts:
#	src/form/Form.jsx
#	test/form/Form.spec.jsx

depoulo reviewed (commented):
react-components doesn't use Prettier (yet?), but you seem to have prettier turned on in your editor, which makes for a very big and hard to understand diff. Can you maybe do a formatting commit first, or tweak your prettier settings to match the existing code style, or something like that?

danilminibaev commented:
@depoulo oh, sure thing, will do, code looks really messy, you right

Commit:
Formatting From.jsx component code

Commit:
Formatting From.jsx component code

- forgot to include arrow-parens rule

depoulo commented on src/form/Form.jsx:
```diff
@@ -79,8 +80,10 @@ export default class Form extends React.Component {
-    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired
-  }
+    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]).isRequired,
+    serverValidationErrors: PropTypes.any,
```
It is not `any`. Please specify a precise `PropTypes.Shape`.

danilminibaev replied:
btw we also have `any` in `values`, should I replace this too?

depoulo replied:
as you wish

danilminibaev replied:
btw, @depoulo, `PropTypes.shape` not always fine for some reason (it throwing warnings), so I used `PropTypes.object`(not sure if its fine though)

depoulo commented on src/form/Form.jsx:
```diff
@@ -164,54 +263,90 @@ export default class Form extends React.Component {
+
+  getFormFieldsStructure = () => {
+    if (this.props.value && this.props.value.size > 0) {
+      return [...this.props.value.keys()]
```
1. Does this mean it doesn't work for nested structures?
2. Do nested structures exist?

danilminibaev replied:
1. yes, it looks so
2. from how we using this it looks like we always expect structures like

```
{
	"formGroup1": {
		'fieldName1': '1', 
		'fieldName2':'a'
	}
}
```

depoulo replied:
ok then

depoulo commented on test/form/Form.spec.jsx:
```diff
@@ -1,53 +1,117 @@
-
-function render (props) {
+import Bluebird from "bluebird";
+import Immutable from "immutable";
```
Why are you adding semicolons all over the place?
Can't you turn of Prettier or

> tweak your prettier settings to match the existing code style

?

danilminibaev replied:
Oops, I realised I updated only the `Form` itself but not the test

depoulo commented on test/form/Form.spec.jsx:
```diff
@@ -1,53 +1,117 @@
+    lastName: "",
     address: {
-      street: ''
+      street: ""
```
Sorry but I'll wait for a more readable diff before continuing, see above comment...

danilminibaev replied:
No probs, my fault

Commit:
Formatting From.spec.jsx test code

depoulo commented on src/form/Form.jsx:
```diff
@@ -164,54 +263,90 @@ export default class Form extends React.Component {
   getError = (name) => {
     return this.state.errors.getIn(parseName(name))
-  }
+  };
```
--- start formatting-only changes ---

depoulo commented on src/form/Form.jsx:
```diff
@@ -164,54 +263,90 @@ export default class Form extends React.Component {
+                .catch(() => {})
+                .then(
+                  () => this.willUnmount || this.setState({ submitting: false })
+                )
```
--- end formatting-only changes ---

depoulo commented on src/form/Form.jsx:
```diff
@@ -164,54 +263,90 @@ export default class Form extends React.Component {
-          submitting: this.state.submitting,
-          reset: () => this.reset()
-        }) : children}
+      <form autoComplete='off' {...other} name={name} onSubmit={this.onSubmit}>
```
--- start formatting-only changes ---

depoulo commented on src/form/Form.jsx:
```diff
@@ -164,54 +263,90 @@ export default class Form extends React.Component {
   static childContextTypes = {
     formValueScope: PropTypes.object
-  }
+  };
```
--- end formatting-only changes ---

depoulo commented on test/form/Form.spec.jsx:
```diff
@@ -20,48 +71,61 @@ function render (props) {
           <div>{`submitting ${submitting}`}</div>
         </div>
-      }
+      )}
```
--- start formatting-only changes ---

depoulo commented on test/form/Form.spec.jsx:
```diff
@@ -269,40 +404,43 @@ describe('Form', function () {
+            onSubmit={() =>
+              this.setState({ unmounted: true }) || Promise.resolve()
+            }
+          >
```
--- end formatting-only changes ---

depoulo commented on test/form/Form.spec.jsx:
```diff
@@ -315,4 +453,100 @@ describe('Form', function () {
+    expect(formComponent, 'to contain', <div>last name server error</div>)
+  })
+
+  it('outputs validation error in popup in case no correspoding filed was found', async function () {
```
field

depoulo reviewed (changes requested):
I wouldn't write about tooltips and popups if it can be anything (rendering stuff and calling a function), but other than that, fine.

I still don't fully understand why you need the ServerErrorMessage component, when in the real usage scenario you can just use the normal ErrorMessage, but maybe you can explain this to me in person...

Commit:
Formatting From.jsx test code parts

Commit:
Formatting From.jsx test code parts pt2

Commit:
Formatting From.jsx and test code parts pt3

Commit:
Added "react-immutable-proptypes"

depoulo commented on src/form/Form.jsx:
```diff
@@ -189,16 +295,18 @@ export default class Form extends React.Component {
 
   render () {
-    const {name, value, onSubmit, onChange, prepare, validate, validateWaitMs, normalize, disabled, children, ...other} = this.props // eslint-disable-line no-unused-vars
+    const {name, value, onSubmit, onChange, prepare, validate, validateWaitMs, normalize, disabled, children, serverValidationErrors, handleUnmappedErrors, ...other} = this.props // eslint-disable-lineno-unused-vars
```
This looks wrong (the `eslint-disable-lineno-unused-vars` part). Wondering why ESLint won't complain.

danilminibaev replied:
oops

depoulo commented on test/form/Form.spec.jsx:
```diff
@@ -315,4 +370,100 @@ describe('Form', function () {
+    })
+  })
+
+  it('removes server side error tooltip when field is changed its content', async function () {
```
Quoting my previous review

> I wouldn't write about tooltips and popups if it can be anything (rendering stuff and calling a function)

danilminibaev replied:
Oh, okay, got it, will remove this

depoulo commented on test/form/Form.spec.jsx:
```diff
@@ -315,4 +370,100 @@ describe('Form', function () {
+    expect(formComponent, 'to contain', <div>last name server error</div>)
+  })
+
+  it('outputs validation error in popup in case no correspoding field was found', async function () {
```
Quoting my previous review

> I wouldn't write about tooltips and popups if it can be anything (rendering stuff and calling a function)

depoulo reviewed (changes requested):
I overheard a review remark by @ingmarh regarding renaming `ServerValidationErrors` to something more generic, because neither are they necessarily validation errors, nor does the code care whether they come from a server, and that made a lot of sense to me as well.

Commit:
Minor changes due to pull request notices

Renamed `serverValidationErrors` property field,
changed description in the test

depoulo committed:
Chore: Whitespace

depoulo committed:
Prettier

Commit:
Fix: Additional check is added in getChangedCompleteErrorList

In case of form value changed but the path for value was not listed in
any of the errors list we want to avoid of using this path
